cmd/kubeadm/app/cmd/token.go

```diff
@@ -255,8 +262,15 @@ func RunGenerateToken(out io.Writer) error {
 	return nil
 }
 
+// BootstrapTokenList represents information for JSON and YAML output produced by 'kubeadm token list'
+type BootstrapTokenList struct {
+	PublicKeyPins        []string                     `json:"publicKeyPins" yaml:"publicKeyPins"`
+	ControlPlaneHostPort string                       `json:"controlPlaneHostPort" yaml:"controlPlaneHostPort"`
+	Tokens               []*kubeadmapi.BootstrapToken `json:"tokens" yaml:"tokens"`
```

Comment:
will do. Please answer my question above regarding copying identical structures. thanks.

Reply:
I did as you've suggested and this conversion code doesn't look good to me to be honest. It creates additional need to update this code when BootstrapToken structures change, which we should avoid in my opinion.
```
                // Convert tokens to the pubic type for marshaling
		var bootstrapTokens []*kubeadmapiv1beta1.BootstrapToken
		for _, token := range tokens {
			bootstrapToken := &kubeadmapiv1beta1.BootstrapToken{
				Token: &kubeadmapiv1beta1.BootstrapTokenString{
					ID:     token.Token.ID,
					Secret: token.Token.Secret,
				},
				Description: token.Description,
				TTL:         token.TTL,
				Expires:     token.Expires,
				Usages:      token.Usages,
				Groups:      token.Groups,
			}
			bootstrapTokens = append(bootstrapTokens, bootstrapToken)
		}
```

Reply:
Just thinking out loud ... JSON/YAML output doesn't contain any info about underlying type, so basically we can use internal BootstrapToken type to avoid this copying. The result would be the same as long as we keep internal and public structures the same. And we'll not need to do this weird copying. WDYT?

Reply:
Sorry for the late response @bart0sh . Yes, this is not the thing to do, the conversion should be done through the API itself. You need something along the lines of this to convert from the internal to the versioned external type:

```
bootstrapToken := &kubeadmapiv1beta1.BootstrapToken{}
kubeadmscheme.Scheme.Convert(token, bootstrapToken, nil)
```

Reply:
going back to the topic of versioning BootstrapTokenList, i'm still on the fence whether we should do that or not.

it's extra maintenance work for us. i think it's best to not do that, and we might break it eventually with an [action required].
i don't think we should consider the output of some of the commands like `token list` and `config images` as API worthy.

Reply:
The likelihood is that if a tool generates `InitConfiguration` with `BootstrapToken`s in it, it might also use this command and the `BootstrapToken`s here will be different (as @bart0sh already found out, `kubeadmapi.BootstrapToken` != `kubeadmapiv1beta1.BootstrapToken`).
The non versioned type is from the internal API and it's unlikely that external users actually import the internal API itself.
Anyway, I think, that we clearly need a policy here and probably hold off merging of this PR until that policy comes into existence.

Reply:
@rosti thank you for the suggestion! kubeadmscheme.Scheme.Convert(token, bootstrapToken, nil) worked just fine for me!

Reply:
@rosti @neolit123 @fabriziopandini as for versioning JSON/YAML output I agree with @neolit123 here. We should try to keep the output stable, but it would not justify an effort to make it versioned in my opinion. At least for now. WDYT?

Reply:
since this change is slightly more complicated. let's hold on it until we merge:
https://github.com/kubernetes/kubernetes/pull/75878

note that PR doesn't use JSON tags at all (so uppercase fields).
but also our existing -o for `kubeadm version` is like that.

ideally this PR should be using the utilities that the other PR adds:
- a common flag option
- a common marshal utility

/hold

Reply:
> note that, that PR doesn't use JSON tags at all (so uppercase fields).

I think it should use them. It would be more risky to remove json field tags from BootstrapToken https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go#L226-L246

Reply:
> I think it should use them

yes, this is settled now based on Version:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/version.go#L36

> It would be more risky to remove json field tags from BootstrapToken 

we cannot because it's a public API object.